pkg/localcontroller/storage/storage.go

```diff
@@ -0,0 +1,148 @@
+	// S3Prefix defines that prefix of url is s3
+	S3Prefix = "s3"
+	// LocalPrefix defines that prefix of url is local host
+	LocalPrefix = ""
+	// S3EndPoint is s3 endpoint of the storage service
+	S3Endpoint = "s3-endpoint"
+	// AccessKeyId is access key id of the storage service
+	AccessKeyID = "ACCESS_KEY_ID"
+	// SecretAccessKey is secret access key of the storage service
+	SecretAccessKey = "SECRET_ACCESS_KEY"
+)
+
+type Storage struct {
+	MinioClient *MinioClient
+}
+
+// Download downloads the file to the local host
+func (s *Storage) Download(objectURL string, localPath string) (string, error) {
+	u, err := url.Parse(objectURL)
+	if err != nil {
+		klog.Errorf("invalid url(%s), error: %v", objectURL, err)
+		return "", fmt.Errorf("invalid url(%s)", objectURL)
+	}
+
+	switch u.Scheme {
+	case LocalPrefix:
+		return s.downloadLocal(objectURL, localPath)
+	case S3Prefix:
+		return s.downloadS3(objectURL, localPath)
+	default:
+		return "", fmt.Errorf("invalid url(%s)", objectURL)
+	}
+}
+
+// downloadLocal copies the local file to another in local host
+func (s *Storage) downloadLocal(objectURL string, localPath string) (string, error) {
+	if !util.IsExists(objectURL) {
+		return "", fmt.Errorf("url(%s) is not exists", objectURL)
+	}
+
+	if localPath == "" {
+		return objectURL, nil
+	}
+
+	dir := path.Dir(localPath)
+	if !util.IsDir(dir) {
+		util.CreateFolder(dir)
+	}
+
+	util.CopyFile(objectURL, localPath)
+
+	return localPath, nil
+}
+
+// downloadS3 downloads the file from url of s3 to the local host
+func (s *Storage) downloadS3(objectURL string, localPath string) (string, error) {
+	if localPath == "" {
+		temporaryDir, err := util.CreateTemporaryDir()
+		if err != nil {
+			return "", err
+		}
+
+		localPath = path.Join(temporaryDir + filepath.Base(objectURL))
+	}
+
+	dir := path.Dir(localPath)
+	if !util.IsDir(dir) {
+		util.CreateFolder(dir)
+	}
+
+	if err := s.MinioClient.downloadFile(objectURL, localPath); err != nil {
+		return "", err
+	}
+
+	return localPath, nil
+}
+
+func (s *Storage) SetCredential(credential interface{}) error {
+	c := credential.(string)
+	m := make(map[string]interface{})
+	if err := json.Unmarshal([]byte(c), &m); err != nil {
+		return err
+	}
+
+	mc := createMinioClient(m[S3Endpoint].(string),
+		m[AccessKeyID].(string),
+		m[SecretAccessKey].(string))
+
+	s.MinioClient = mc
+
+	return nil
+}
+
+// Upload uploads the file in local host to another(e.g., "s3")
+func (s *Storage) Upload(localPath string, objectURL string) (string, error) {
+	u, err := url.Parse(objectURL)
+	if err != nil {
+		klog.Errorf("invalid url(%s), error: %v", objectURL, err)
+		return "", fmt.Errorf("invalid url(%s)", objectURL)
+	}
+
+	switch u.Scheme {
+	case S3Prefix:
+		return s.uploadS3(localPath, objectURL)
+	default:
+		return "", fmt.Errorf("invalid url(%s)", objectURL)
+	}
+}
+
+// uploadS3 uploads the file in local host to the url of s3
+func (s *Storage) uploadS3(localPath string, objectURL string) (string, error) {
```

Comment:
nits: just return error, since the string return value is same the objectURL input param

Reply:
fix it